File: memetool_sol.py
from solana.rpc.api import Client
import time
from datetime import datetime
import json
import requests

class SolanaMemeMonitor:
    def __init__(self):
        self.solana_client = Client("https://api.mainnet-beta.solana.com")
        self.known_tokens = set()
        self.target_token_name = "GAIM Studio"
        
    def is_target_token(self, token_info):
        if not token_info or 'name' not in token_info:
            return False
        
        token_name = token_info['name']
        return token_name == self.target_token_name

    def get_token_info(self, token_address, max_retries=3):
        for attempt in range(max_retries):
            try:
                # 添加请求头，模拟浏览器行为
                headers = {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
                }
                
                # 获取基本代币信息
                response = requests.get(
                    f"https://public-api.solscan.io/token/meta/{token_address}",
                    headers=headers,
                    timeout=10
                )
                if response.status_code == 200:
                    token_data = response.json()
                    if self.is_target_token(token_data):
                        # 获取代币创建者信息
                        creator_response = requests.get(
                            f"https://public-api.solscan.io/token/holders/{token_address}",
                            headers=headers,
                            timeout=10
                        )
                        if creator_response.status_code == 200:
                            creator_data = creator_response.json()
                            if creator_data.get('data'):
                                token_data['creator'] = creator_data['data'][0].get('owner', '未知')
                        return token_data
                return None
            except requests.exceptions.RequestException as e:
                if attempt == max_retries - 1:  # 最后一次重试
                    print(f"获取代币信息错误 (尝试 {attempt + 1}/{max_retries}): {str(e)}")
                else:
                    print(f"重试获取代币信息 (尝试 {attempt + 1}/{max_retries})...")
                    time.sleep(2)  # 重试前等待2秒
            except Exception as e:
                print(f"获取代币信息时发生未知错误: {str(e)}")
                if attempt < max_retries - 1:
                    time.sleep(2)
        return None

    def get_new_tokens(self, max_retries=3):
        for attempt in range(max_retries):
            try:
                headers = {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
                }
                response = requests.get(
                    "https://token.jup.ag/all",
                    headers=headers,
                    timeout=10
                )
                if response.status_code == 200:
                    tokens = response.json()
                    return [token['address'] for token in tokens]
                return []
            except requests.exceptions.RequestException as e:
                if attempt == max_retries - 1:
                    print(f"获取新代币列表错误 (尝试 {attempt + 1}/{max_retries}): {str(e)}")
                else:
                    print(f"重试获取代币列表 (尝试 {attempt + 1}/{max_retries})...")
                    time.sleep(2)
            except Exception as e:
                print(f"获取代币列表时发生未知错误: {str(e)}")
                if attempt < max_retries - 1:
                    time.sleep(2)
        return []

    def monitor_new_tokens(self):
        print(f"开始监测名称为 '{self.target_token_name}' 的新代币...")
        while True:
            try:
                new_tokens = self.get_new_tokens()
                
                for token_address in new_tokens:
                    if token_address not in self.known_tokens:
                        self.known_tokens.add(token_address)
                        token_info = self.get_token_info(token_address)
                        if token_info:
                            print("\n" + "="*50)
                            print(f"发现目标代币！")
                            print(f"时间: {datetime.now()}")
                            print(f"代币名称: {token_info.get('name')}")
                            print(f"代币地址: {token_address}")
                            print(f"代币符号: {token_info.get('symbol')}")
                            print(f"创建者地址: {token_info.get('creator', '未知')}")
                            print("="*50 + "\n")
                
                time.sleep(5)
                
            except Exception as e:
                print(f"监控过程中出错: {str(e)}")
                time.sleep(5)

if __name__ == "__main__":
    monitor = SolanaMemeMonitor()
    monitor.monitor_new_tokens()

File: memetool_sol.js
const fs = require('fs');
const path = require('path');
const axios = require('axios');


class SolanaTokenMonitor {
  constructor(options = {}) {
    this.knownTokensFile = options.knownTokensFile || 'known_tokens.json';
    this.alertLogFile = options.alertLogFile || 'token_alerts.json';
    this.targetTokenName = options.targetTokenName || 'Replicat-One';
    this.intervalMs = options.intervalMs || 10_000; 
    this.knownTokens = new Set(this.loadKnownTokens());
  }

  loadKnownTokens() {
    try {
      const filePath = path.resolve(__dirname, this.knownTokensFile);
      if (fs.existsSync(filePath)) {
        const raw = fs.readFileSync(filePath, 'utf-8');
        const arr = JSON.parse(raw);
        if (Array.isArray(arr)) {
          return arr;
        }
      }
    } catch (err) {
      console.error(`加载 knownTokens 文件出错: ${err.message}`);
    }
    return [];
  }

  saveKnownTokens() {
    try {
      const filePath = path.resolve(__dirname, this.knownTokensFile);
      fs.writeFileSync(filePath, JSON.stringify(Array.from(this.knownTokens), null, 2));
    } catch (err) {
      console.error(`保存 knownTokens 文件出错: ${err.message}`);
    }
  }

  async getNewTokens() {
    try {
      const response = await axios.get('https://token.jup.ag/all', {
        headers: {
          'User-Agent': 'Mozilla/5.0',
          'Accept': 'application/json'
        },
        timeout: 10_000 
      });

      const data = response.data || [];
      if (!Array.isArray(data)) {
        console.error('Jupiter 返回数据格式异常');
        return [];
      }
      return data.map(token => token.address).filter(Boolean);
    } catch (error) {
      console.error(`获取新代币列表错误: ${error.message}`);
      return [];
    }
  }

  async getTokenInfo(tokenAddress) {
    for (let attempt = 1; attempt <= 3; attempt++) {
      try {
        const response = await axios.get(`https://public-api.solscan.io/token/meta/${tokenAddress}`, {
          headers: { 'User-Agent': 'Mozilla/5.0', 'Accept': 'application/json' },
          timeout: 5_000
        });

        if (response.status === 200 && this.isTargetToken(response.data)) {
          const creatorInfo = await this.getTokenCreator(tokenAddress);
          response.data.creator = creatorInfo || '未知';
          return response.data;
        }
        return null;
      } catch (error) {
        if (error.response && error.response.status === 404) {
          return null;
        }
        console.error(`第 ${attempt} 次获取代币[${tokenAddress}]信息失败: ${error.message}`);

        if (attempt === 3) {
          return null;
        }
      }
    }
    return null;
  }

  async getTokenCreator(tokenAddress) {
    try {
      const response = await axios.get(
        `https://public-api.solscan.io/token/holders/${tokenAddress}`,
        {
          headers: { 'User-Agent': 'Mozilla/5.0', 'Accept': 'application/json' },
          timeout: 5_000
        }
      );
      if (response.status === 200 && response.data.data && response.data.data.length > 0) {
        return response.data.data[0].owner || '未知';
      }
      return null;
    } catch (err) {
      console.log(`获取创建者信息失败: ${err.message}`);
      return null;
    }
  }

  isTargetToken(tokenInfo) {
    if (!tokenInfo || !tokenInfo.name) return false;
    const name = tokenInfo.name.trim().toLowerCase();
    const target = this.targetTokenName.trim().toLowerCase();
    return name === target;
  }

  logTokenInfo(tokenInfo, tokenAddress) {
    const separator = '='.repeat(50);
    console.log(`\n${separator}`);
    console.log('发现目标代币！');
    console.log(`时间: ${new Date().toLocaleString()}`);
    console.log(`代币名称: ${tokenInfo.name}`);
    console.log(`代币地址: ${tokenAddress}`);
    console.log(`代币符号: ${tokenInfo.symbol}`);
    console.log(`创建者地址: ${tokenInfo.creator || '未知'}`);
    console.log(`${separator}\n`);
  }

  saveAlertToFile(tokenInfo, tokenAddress) {
    try {
      const data = {
        timestamp: new Date().toISOString(),
        tokenName: tokenInfo.name,
        tokenAddress: tokenAddress,
        symbol: tokenInfo.symbol,
        creator: tokenInfo.creator
      };
      const filePath = path.resolve(__dirname, this.alertLogFile);
      fs.appendFileSync(filePath, JSON.stringify(data) + '\n');
    } catch (err) {
      console.error(`保存到文件时出错: ${err.message}`);
    }
  }

  async monitorTokensLoop() {
    console.log(`开始监测名称为 '${this.targetTokenName}' 的新代币...`);

    while (true) {
      try {
        const newTokens = await this.getNewTokens();
        const tasks = [];
        for (const tokenAddress of newTokens) {
          if (!this.knownTokens.has(tokenAddress)) {
            tasks.push(
              async () => {
                this.knownTokens.add(tokenAddress);
                const tokenInfo = await this.getTokenInfo(tokenAddress);
                if (tokenInfo) {
                  this.logTokenInfo(tokenInfo, tokenAddress);
                  this.saveAlertToFile(tokenInfo, tokenAddress);
                }
              }
            );
          }
        }
        await Promise.all(tasks);

        this.saveKnownTokens();
      } catch (error) {
        console.error(`监控过程中出错: ${error.message}`);
      }
      await this.sleep(this.intervalMs);
    }
  }

  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

(async () => {
  const monitor = new SolanaTokenMonitor({
    targetTokenName: 'Replicat-One',
    knownTokensFile: 'known_tokens.json',
    alertLogFile: 'token_alerts.json',
    intervalMs: 10_000
  });
  await monitor.monitorTokensLoop();
})();

